docs/storage/storage.md

```diff
@@ -0,0 +1,186 @@
+tree is mutated.
+
+### NodeIDs
+
+Nodes in storage are uniquely identified by a `NodeID`. This combines a tree path with
+a revision number. The path is effectively used as a bitwise subtree prefix in the tree. 
+
+The same `NodeID` objects are used by both logs and maps but they are interpreted differently.
+There are API functions that create them for each case. Mixing node ID types in API calls will
+give incorrect results.
+
+### Subtree Compaction
+
+As an optimization the tree is not stored as a set of raw nodes but at as a collection of subtrees.
+
+Currently, subtrees must be be a multiple of 8 levels deep (referred to as `strataDepth` in the
+code). Only the bottom level nodes the "leaves" of the subtree are physically stored.
+Intermediate subtree nodes are rehashed from the "leaves" when the subtree is loaded into memory.
+
+Subtrees are keyed by the `NodeID` of the root (effectively a path prefix) and contain the
+intermediate nodes for that subtree, identified by their suffix. These are actually stored in a
+proto map where the key is the suffix path. 
+
+Subtrees are versioned to support the access model for nodes described above. Node addresses within
+the model are distinct because the path to a subtree must be unique.
+
+When node updates are applied they will affect one or more subtrees and caching is used to increase
+efficiency. After all updates have been done in-memory the cache is flushed to storage so each
+affected subtree is only written once.
+
+Subtrees are helpful for reads because it is likely that many of the nodes traversed in
+Merkle paths for proofs are part of the same subtree. The number of subtrees involved in a path
+through a large tree from the leaves to the root is also bounded. For writes the subtree update
+batches what would be many smaller writes into one larger but manageable one.
+
+We gain space efficiency by not storing intermediate nodes. This is a big saving as it avoids
+storing entire tree levels, which get very large as the tree grows. This is magnified further as we
+store many versions of the tree.
+
+#### Subtree Diagrams
+
+This diagram shows a tree as it might actually be stored by our code using subtrees of depth 8.
+
+Each subtree does not include its "root" node, though this counts as part of the depth. There are
+additional subtrees below and to the right of the child subtree shown, they can't easily be shown
+in the diagram. Obviously, there could be less than 256 "leaf" nodes in the subtrees if they are not
+yet fully populated.
+
+![strata depth 8 tree](StratumDepth8.png, "Stratum Depth 8")
+
+As it's hard to visualize the structure at scale with stratum depth 8 some examples of smaller
+depths might make things clearer, though these are not supported by the code the diagrams are
+much simpler.
+
+This diagram shows a tree with strata depth 2. It is a somewhat special case as all the levels are
+stored. Note that the root node is never stored and is always recalculated.
```

Comment:
So another surprise here for me now I read this and see the diagrams:  you say that StratumDepth2.png shows a subtree of stratum depth=2, yet there are only the leaf nodes in blue inside the wedge.  So you include the subtree root when counting depth.

So a general property of all these things is that, despite the root not being stored - you say it is always recalculated - it must correspond to one of the 'leaves' of the next stratum (I mean subtree) above and in those it will be stored.

i.e in all the diagrams the blue node at a sub-tree root 'overlays' a yellow node from the parent layer.